ls1intum/Artemis
# `General`: Move LTI custom parameters to OnlineCourseConfiguration

### Checklist
#### General


#### Server
#### Client

### Motivation and Context

The custom parameters for LTI launches are complicated to configure, and most importantly put the responsibility of configuring it correctly on the Platform (e.g. Moodle), instead of on the Artemis course being used as an online course. We want the configuration to be done on Artemis, instead of making/allowing instructors of the platforms to do the configuration.

There were two custom parameters, `requireExistingUser` and `lookupUserByEmail`. `requireExistingUser` defines whether we allow users to be created as part of an LTI launch, or if we only want already existing Artemis users to be able to launch those LTI exercises. `lookupUserByEmail` defines whether we check if there's already an Artemis user with the email provided in the launch. Both parameters could be configured on the platform and be different for each exercise. If no parameters were provided, both values would default to false.

### Description

In this PR we change these custom parameters so that:
-  `requireExistingUser` - we add this as a new OnlineCourseConfiguration field. For each online course, the instructor can configure whether new users are allowed, or if only existing users can launch LTI exercises of that course. The default is false (meaning new users are allowed). The behavior of the parameter stays the same.
-  `lookupUserByEmail` - we remove the parameter, and default the behavior to true. We **always** want to check if there's already a user in Artemis with the same email, otherwise we will create a user with a duplicate email (or get an error when we try to do so), which is undesired behavior

### Steps for Testing
Prerequisites:
- 1 Instructor in Artemis
- 1 Instructor in https://moodle.ase.in.tum.de/

#### TestrequireExistingUser:

1. Log in to Artemis as instructor
2. Create/edit a course to be an onlineCourse 
3. Verify setting/unsetting the requireExistingUser checkbox works and the state is saved with the course
4. Login to Moodle as instructor
5. Setup LTI1.0 or LTI1.3 exercise (you don't need to set custom parameters anymore)
6. Launch LTI exercise with a new user with requireExistingUser = true and verify that you get an error
7. Launch LTI exercise with a new user with requireExistingUser = false and verify that you don't get an error
8. Launch LTI exercise with an existing user with requireExistingUser = true and verify that you don't get an error

#### Test lookupUserByEmail:

1. Log in to Artemis as instructor
2. Create/edit a course to be an onlineCourse 
3. Login to Moodle as instructor
4. Setup LTI1.0 or LTI1.3 exercise (you don't need to set custom parameters anymore)
5. Create a user in Moodle with an email matching the email of an existing user in Artemis
6. Login in Moodle as the user created in the step before
7. Launch the LTI exercise and verify you were matched to the user with the same email in Artemis

#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| LtiService.java | % | % |
| Lti10Service.java | % | % |
| Lti13Service.java | % | % |

### Screenshots
There is just one added checkbox when creating/editing a course for setting the "requireExistingUser" field in the OnlineCourseConfiguration. This will be improved/reworked in the PR to improve the LTI configuration in the UI: https://github.com/ls1intum/Artemis/pull/5812

![image](https://user-images.githubusercontent.com/11006274/201475312-09a3d636-4d5e-4c94-85fe-c6fef1e7cee5.png)
